compiler/rustc_lint/src/unused.rs

```diff
@@ -1038,7 +1038,20 @@ impl EarlyLintPass for UnusedParens {
         if let ast::TyKind::Paren(r) = &ty.kind {
             match &r.kind {
                 ast::TyKind::TraitObject(..) => {}
-                ast::TyKind::BareFn(b) if b.generic_params.len() > 0 => {}
+                ast::TyKind::BareFn(b) => {
+                    let spans = if let Some(r) = r.span.find_ancestor_inside(ty.span) {
+                        Some((ty.span.with_hi(r.lo()), ty.span.with_lo(r.hi())))
+                    } else {
+                        None
+                    };
+                    let sm = cx.sess().source_map();
+                    if let Some((_, r)) = spans &&
+                        b.generic_params.len() > 0 &&
+                        sm.span_followed_by(r, ":") {
```

Comment:
This aims to check that we are in a where predicate, doesn't it?
Could you prefer a more explicit check, with a new `check_where_predicate` method or something like that?

Reply:
You means add `check_where_predicate` in `EarlyLintPass`?
https://github.com/rust-lang/rust/blob/f206533fd40378da6e2a07567e8d7592edd13ee4/compiler/rustc_lint/src/early.rs#L249

Seems `walk_where_predicate` again will also call to `visit_ty`, then call to `check_ty`:

https://github.com/rust-lang/rust/blob/f206533fd40378da6e2a07567e8d7592edd13ee4/compiler/rustc_ast/src/visit.rs#L623

Reply:
> You means add `check_where_predicate` in `EarlyLintPass`?
> 

Yes.

> https://github.com/rust-lang/rust/blob/f206533fd40378da6e2a07567e8d7592edd13ee4/compiler/rustc_lint/src/early.rs#L249
> 
> Seems `walk_where_predicate` again will also call to `visit_ty`, then call to `check_ty`:
> 
> https://github.com/rust-lang/rust/blob/f206533fd40378da6e2a07567e8d7592edd13ee4/compiler/rustc_ast/src/visit.rs#L623

You'll need to ensure that we don't lint on that case from `check_ty`, but from `check_where_predicate`.

Reply:
How can we customize the behavior of `visit_where_predicate` for an specific lint? In this case, it's the lint for `UnusedParens`.

Reply:
Seems we can not add a return value for `check_where_predicate`.